### PR TITLE
Async Posting: Show notices for failed uploads

### PR DESCRIPTION
**Pull request by frosty**
Partial implementation of #8713. This PR follows on from #8921, by adding notices for failure cases when either media or posts fail to upload.

Note that the messages differ slightly from the designs, as the circumstances under which uploads can fail are I think a little different to what we first thought. There are two cases in which these failures might be displayed:

1. Media for a post fails to upload. In this case, we haven't yet tried to upload the post, so we can just display a normal media failure notice (except we mention that the post has failed). The retry button on the notice will retry the uploads, which will still be observed by the `PostCoordinator`. If the media uploads then complete, the post coordinator will upload the post as normal.

img_2017

2. If the media has uploaded, but the post itself fails to upload, we'll display a new variation on the Post upload notice:

img_2021

The Retry button will simply retry the upload through the `PostCoordinator`.

**To test:**

* To test both of these cases, it's easiest to turn on the Very Bad Network link conditioner, so that you have more time to perform actions before the posts complete.
* For notice 1, you can do the following:
  * Open the editor, and add a couple of media items
  * Before they finish uploading, use the Async Publish button
  * Switch airplane mode on, and the notice should appear. Quickly switch airplane mode off and then tap the Retry button to test that. You could increase the hide duration in NoticePresenter if you need more time.
* For notice 2, you can do the following:
  * Open the editor, and don't add any media items
  * Use the Async Publish button
  * Switch airplane mode on before the publish completes. You should see the notice. Quickly switch airplane mode off and then tap the Retry button to test that. You could increase the hide duration in NoticePresenter if you need more time.

### CODE REVIEW

File: WordPress/Classes/Services/PostCoordinator.swift [hunk 1]
```diff
@@ -72,4 +86,8 @@ class PostCoordinator: NSObject {
 
         post.content = postContent
     }
+
+    private func removeObserver(for post: AbstractPost) {
+        observerUUIDs.removeValue(forKey: post)
```
Comment (SergioEstevao) [inline WordPress/Classes/Services/PostCoordinator.swift]: Shouldn't we call: `mediaCoordinator.removeObserver()`

File: WordPress/Classes/Services/PostCoordinator.swift [hunk 2]
```diff
@@ -23,18 +25,27 @@ class PostCoordinator: NSObject {
...
                 switch state {
                 case .ended:
                     self.updateReferences(to: media, in: post)
+
                     // Let's check if media uploading is still going, if all finished with success then we can upload the post
                     if !mediaCoordinator.isUploadingMedia(for: post) && !post.hasFailedMedia {
+                        self.removeObserver(for: post)
                         self.upload(post: post)
                     }
                 default:
```
Comment (SergioEstevao) [inline WordPress/Classes/Services/PostCoordinator.swift]: We need to add a case where media the last media for the posts ends up by failing. Something like this:
```               
case  .failed(_):
                    if !mediaCoordinator.isUploadingMedia(for: post) {
                        self.removeObserver(for:post)                    
                    }
```

Comment (frosty) [reply WordPress/Classes/Services/PostCoordinator.swift]: We don't want to remove the observer in the case where all of the media has failed – in this situation, the media coordinator will display a Notice to the user, giving them the chance to retry the uploads. If we leave the observer in place here, then the user can retry the uploads and the observer block will pick it up and upload the post afterwards.

Comment (SergioEstevao) [reply WordPress/Classes/Services/PostCoordinator.swift]: Humm fair point, let's go forward like this for now, I will check on my further work for post upload cancel if all is good.

Comment (frosty) [reply WordPress/Classes/Services/PostCoordinator.swift]: I think you're right, there is a situation where the observer could be left hanging, but I'm not sure the right point to detect that and remove it. Here's one scenario:

* A user starts publishing a post that has media to upload. We attach an observer.
* The media fails, but the user chooses not to restart it. They go into the post and remove the media.
* They then try again to publish the post. The post publishes successfully, but we never removed the observer.

**Pull request merged**